File: src/main/scala/firelib/robot/Frequencer.cs
using System;
using System.Collections.Generic;
using System.Threading;
using QuantLib.Domain;

namespace Fire.Common.Robot
{
    public class Frequencer
    {
        private readonly Interval interval;

        private readonly List<Action<DateTime>> timeListeners = new List<Action<DateTime>>();

        private Timer timer;

        private long lastTimeTrigger;

        private long precision = 250;

        public void AddListener(Action<DateTime> act)
        {
            lock (timeListeners)
            {
                timeListeners.Add(act);
            }
        }

        private void CheckTime()
        {            
            long epochTick = DateTime.Now.Ticks;
            long rounded = (epochTick/interval.Ticks)*interval.Ticks;
            if (lastTimeTrigger != rounded)
            {
                lastTimeTrigger = rounded;
                NotifyListeners(new DateTime(rounded));                
            }

        }

        public void Start()
        {
            timer = new Timer((o) => CheckTime(), null, precision, precision);
        }

        private void NotifyListeners(DateTime ctime)
        {
            lock (timeListeners)
            {
                foreach (var timeListener in timeListeners)
                {
                    timeListener(ctime);
                }
            }
        }

        public Frequencer(Interval interval)
        {
            this.interval = interval;
        }
    }
}

File: src/main/scala/firelib/robot/Frequencer.scala
package firelib.robot

import java.util.concurrent.{CopyOnWriteArrayList, Executors, TimeUnit}

import firelib.domain.Interval
import org.joda.time.DateTime

import scala.collection.JavaConversions._

class Frequencer(val interval: Interval, val precisionMs: Long = 100) extends Runnable {

    private val timeListeners = new CopyOnWriteArrayList[DateTime => Unit]()

    private var lastTimeTrigger: Long = _

    val executor = Executors.newSingleThreadScheduledExecutor()

    def AddListener(act: DateTime => Unit) = {
        timeListeners += act
    }

    def Start() = {
        executor.scheduleAtFixedRate(this, precisionMs, precisionMs, TimeUnit.MILLISECONDS)
    }

    private def NotifyListeners(ctime: DateTime) = for (tl <- timeListeners) tl(ctime)

    override def run(): Unit = {
        val epochTick = System.currentTimeMillis();
        val rounded = (epochTick / interval.durationMs) * interval.durationMs;
        if (lastTimeTrigger != rounded) {
            lastTimeTrigger = rounded;
            NotifyListeners(new DateTime(rounded));
        }
    }
}
